kratos/processes/find_global_nodal_neighbours_for_conditions_process.h

```diff
@@ -0,0 +1,117 @@
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_FIND_GLOBAL_NODAL_NEIGHBOURS_FOR_CONDITIONS_PROCESS_H_INCLUDED)
+#define KRATOS_FIND_GLOBAL_NODAL_NEIGHBOURS_FOR_CONDITIONS_PROCESS_H_INCLUDED
+
+// System includes
+#include <iostream>
+#include <string>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/global_pointer_variables.h"
+#include "includes/model_part.h"
+#include "processes/find_global_nodal_neighbours_for_entities_process.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+ */
+class FindGlobalNodalNeighboursForConditionsProcess
+    : public FindNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of FindGlobalNodalNeighboursForConditionsProcess
+    KRATOS_CLASS_POINTER_DEFINITION(FindGlobalNodalNeighboursForConditionsProcess);
+
+    using BaseType =
+        FindNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType>;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// constructor.
+    FindGlobalNodalNeighboursForConditionsProcess(
+        const DataCommunicator& rDataCommunicator,
+        ModelPart& rModelPart)
+        : BaseType(rDataCommunicator, rModelPart, NEIGHBOUR_CONDITION_NODES)
+    {
+    }
+
+    /// Destructor.
+    ~FindGlobalNodalNeighboursForConditionsProcess() override = default;
+
+    /// Assignment operator.
+    FindGlobalNodalNeighboursForConditionsProcess& operator=(
+        FindGlobalNodalNeighboursForConditionsProcess const& rOther) = delete;
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        return "FindGlobalNodalNeighboursForConditionsProcess";
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "FindGlobalNodalNeighboursForConditionsProcess";
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const override
+    {
+    }
+
+    ///@}
+
+}; // Class FindGlobalNodalNeighboursForConditionsProcess
+
+///@}
+///@name Input and output
+///@{
+
+/// input stream function
+inline std::istream& operator>>(
+    std::istream& rIStream,
+    FindGlobalNodalNeighboursForConditionsProcess& rThis);
+
+/// output stream function
+inline std::ostream& operator<<(
+    std::ostream& rOStream,
+    const FindGlobalNodalNeighboursForConditionsProcess& rThis)
+{
+    rThis.PrintInfo(rOStream);
+    rOStream << std::endl;
+    rThis.PrintData(rOStream);
+
+    return rOStream;
+}
+///@}
+
+} // namespace Kratos.
+
+#endif // KRATOS_FIND_GLOBAL_NODAL_NEIGHBOURS_FOR_CONDITIONS_PROCESS_H_INCLUDED defined
```

Comment:
As I understand this class does not add any functionality, probably is best if we just define

```C++
typename FindNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType> FindNodalNeighboursForConditionsProcess;

```

Where it needs to be used. Same with the

Reply:
thats true... I will change it :)

Reply:
I am afraid, Its little bit harder to remove. In FindNodalNeighboursForEntitiesProcess i have the constructor with GlobalPointersVector type variable. GlobalPointersVector type variables are not exposed to python, therefore I cannot expose FindNodalNeighboursForEntitiesProcess to python. Do you know a way to give ctors predefined arguments when exposing to python (it cannot be a py::args since user cannot specify it from python). :/

Reply:
I think I know how to do it, let me try a couple of things.

Reply:
Ok, seems it can be done with some gymnastics:

A quick note first: I would leave the exit variable type as something that in principle one cannot set from python (but is easy to extend overloading the init function in the new export function)

Since pybind uses the templates of the init<> method to deduce the number of arguments, it will automatically kill any default value, but since those are only syntax suggar you can mimic the same behavior by exposing overloaded constructors with the non default variables which calls the constructor with the default value. For example:

```C++
    typedef FindGlobalNodalNeighboursForEntitiesProcess<ModelPart::ElementsContainerType> FindGlobalNodalNeighboursProcess;
    py::class_<FindGlobalNodalNeighboursProcess, FindGlobalNodalNeighboursProcess::Pointer, Process>
        (m,"FindGlobalNodalNeighboursProcess")
    .def(py::init([](const DataCommunicator& rDataComm, ModelPart& rModelPart) {
        return std::unique_ptr<FindGlobalNodalNeighboursProcess>(new FindGlobalNodalNeighboursProcess(rDataComm, rModelPart, NEIGHBOUR_NODES));
    }))
    .def("ClearNeighbours",&FindGlobalNodalNeighboursProcess::ClearNeighbours)
    .def("GetNeighbourIds",&FindGlobalNodalNeighboursProcess::GetNeighbourIds)
    ;

    typedef FindGlobalNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType> FindGlobalNodalNeighboursForConditionsProcess;
    py::class_<FindGlobalNodalNeighboursForConditionsProcess, FindGlobalNodalNeighboursForConditionsProcess::Pointer, Process>
        (m,"FindGlobalNodalNeighboursForConditionsProcess")
    .def(py::init([](const DataCommunicator& rDataComm, ModelPart& rModelPart) {
        return std::unique_ptr<FindGlobalNodalNeighboursForConditionsProcess>(new FindGlobalNodalNeighboursForConditionsProcess(rDataComm, rModelPart, NEIGHBOUR_CONDITION_NODES));
    }))
    .def("ClearNeighbours",&FindGlobalNodalNeighboursForConditionsProcess::ClearNeighbours)
    .def("GetNeighbourIds",&FindGlobalNodalNeighboursForConditionsProcess::GetNeighbourIds)
    ;
```

Also notice that, since this works, I think we could remove all the auxiliary classes, leave only the *.cpp for the container specializations (and leave the element ones as default) and set that variable as default in the constructor of the main class in C++. This way the the code remains exactly the same but with the option to overloaded both in terms of the variable type and the container for which you perform the search: You could even bind one to the other:

```C++
template <class TContainerType = ModelPart::ElementsContainerType>
    static Variable<GlobalPointersVector<NodeType>> DefaultContainer;

    /// Constructor
    FindGlobalNodalNeighboursForEntitiesProcess(
        const DataCommunicator& rDataCommunicator,
        ModelPart& rModelPart,
        const Variable<GlobalPointersVector<NodeType>>& rOutputVariable = FindGlobalNodalNeighboursForEntitiesProcess::DefaultContainer)
        : mrModelPart(rModelPart),
          mrDataCommunicator(rDataCommunicator),
          mrOutputVariable(rOutputVariable)
    {
    }
```
And then outside the class:

```C++
template class FindGlobalNodalNeighboursForEntitiesProcess<ModelPart::ElementsContainerType>::DefaultContainer = NEIGHBOUR_NODES;
template class FindGlobalNodalNeighboursForEntitiesProcess<ModelPart::ConditionsContainerType>::;DefaultContainer = NEIGHBOUR_CONDITION_NODES
```

This also would get rid of the exposing problem (but it would be nice to leave the overloaded version so the user can select the container).

What do you think. Also probably comments from @RiccardoRossi as he did the class after all.

Reply:
I like the first idea (seems to be easy to read and understand from first looks). For the time being I will change it like that, and will wait for @RiccardoRossi comments as well :)

Reply:
done :) @roigcarlo thanks for the detailed explanation :) :+1: